### ISSUE TITLE
[Bug?] Large (font size) merged fonts leading to blocky, fat text

### ISSUE THREAD

**Issue by DubbleClick** [ocornut/imgui#7777]
### Version/Branch of Dear ImGui:

Version 1.90.9, Branch: master/docking
Happens in other versions since at least 1.80 too.

### Back-ends:

imgui_impl_dx9.cpp + imgui_impl_win32.cpp

### Compiler, OS:

MSVC 1940, v143, Windows 10/11

### Full config/build information:

```
Dear ImGui 1.90.9 (19090)
--------------------------------
sizeof(size_t): 4, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 24
define: __cplusplus=199711
define: IMGUI_DISABLE_OBSOLETE_FUNCTIONS
define: IMGUI_DISABLE_OBSOLETE_KEYIO
define: IMGUI_USE_BGRA_PACKED_COLOR
define: _WIN32
define: _MSC_VER=1940
define: _MSVC_LANG=202004
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
--------------------------------
io.BackendPlatformName: imgui_impl_win32
io.BackendRendererName: imgui_impl_dx9
io.ConfigFlags: 0x00000028
 NavNoCaptureKeyboard
 NoMouseCursorChange
io.ConfigViewportsNoDecoration
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x0000000E
 HasMouseCursors
 HasSetMousePos
 RendererHasVtxOffset
--------------------------------
io.Fonts: 6 fonts, Flags: 0x00000000, TexSize: 4096,16384
io.DisplaySize: 2560.00,1440.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------
style.WindowPadding: 8.00,8.00
style.WindowBorderSize: 0.00
style.FramePadding: 4.00,3.00
style.FrameRounding: 2.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 8.00,4.00
style.ItemInnerSpacing: 4.00,4.00
```

### Details:

When adding fonts in multiple sizes with multiple glyph ranges, fonts of a font atlas occasionally end up very blocky - see attached screenshot.

Attempts at fixing this include the following: creating multiple font atlases and specifying atlas->TexDesiredWidth. When it happens, the problem is isolated to a font atlas - others are not affected.

It is more prone to happen when adding multiple fonts of large font size (40/48pt). Any idea what may cause this or how to tackle it? The proper way to introduce even more sizes in the future would likely be to use font scaling instead, but I would like to understand what causes this, in order to avoid the issue.

https://github.com/gwdevhub/GWToolboxpp/compare/dev...bugged_fonts

### Screenshots/Video:

Blocky:
![image](https://github.com/ocornut/imgui/assets/7896469/694da66f-005c-48e2-b2e6-556df8efd6a2)
Normal:
![image](https://github.com/ocornut/imgui/assets/7896469/6cba290a-d1d0-4f93-9e68-11b1096577fb)

### Minimal, Complete and Verifiable Example code:

```cpp
// loading fonts
const auto& io = ImGui::GetIO();

const std::vector<std::pair<const wchar_t*, const ImWchar*>> fonts_on_disk = {
    {L"Font.ttf", io.Fonts->GetGlyphRangesDefault()},
    {L"Font_Japanese.ttf", io.Fonts->GetGlyphRangesJapanese()},
    {L"Font_Cyrillic.ttf", io.Fonts->GetGlyphRangesCyrillic()},
    {L"Font_ChineseTraditional.ttf", io.Fonts->GetGlyphRangesChineseFull()},
    {L"Font_Korean.ttf", io.Fonts->GetGlyphRangesKorean()}
};

struct FontData {
    const ImWchar* glyph_ranges = nullptr;
    size_t data_size = 0;
    void* data = nullptr;
};
std::vector<FontData> fonts;
for (size_t i = 0; i < fonts_on_disk.size(); i++) {
    const auto& [font_name, glyph_range] = fonts_on_disk[i];
    const utf8::string utf8path = Resources::GetPathUtf8(font_name);
    size_t size;
    void* data = ImFileLoadToMemory(utf8path.bytes, "rb", &size, 0);
    if (data) {
        fonts.push_back({glyph_range, size, data});
    }
    else if (i == 0) {
        // first one cannot fail
        printf("Failed to find Font.ttf file\n");
        fonts_loaded = true;
        fonts_loading = false;
        return;
    }
}
static constexpr std::array<ImWchar, 3> fontawesome5_glyph_ranges = {ICON_MIN_FA, ICON_MAX_FA, 0};

auto cfg = ImFontConfig();
cfg.PixelSnapH = true;
cfg.FontDataOwnedByAtlas = true;
cfg.OversampleH = 2;
cfg.OversampleV = 1;

auto add_font_set = [&cfg, &fonts](ImFontAtlas* atlas, const float size, ImFont*& font_ptr, const bool add_all = true) {
    cfg.MergeMode = false;
    for (const auto& font : fonts) {
        atlas->AddFontFromMemoryTTF(font.data, font.data_size, size, &cfg, font.glyph_ranges);
        cfg.MergeMode = true; // for all but the first
        if (!add_all) break;
    }
    atlas->AddFontFromMemoryCompressedTTF(
        fontawesome5_compressed_data, fontawesome5_compressed_size, size, &cfg, fontawesome5_glyph_ranges.data());
    font_ptr = atlas->Fonts.back();
    cfg.FontDataOwnedByAtlas = false;
};

add_font_set(io.Fonts, static_cast<float>(FontSize::text), font_text); // 16.f
add_font_set(io.Fonts, static_cast<float>(FontSize::header2), font_header2); // 18.f
add_font_set(io.Fonts, static_cast<float>(FontSize::header1), font_header1); // 20.f

cfg.OversampleH = 1;
add_font_set(io.Fonts, static_cast<float>(FontSize::widget_label), font_widget_label); // 24.f
add_font_set(io.Fonts, static_cast<float>(FontSize::widget_small), font_widget_small); // 40.f - passing false as add_all means no problem
add_font_set(io.Fonts, static_cast<float>(FontSize::widget_large), font_widget_large); // 48.f - passing false as add_all means no problem

if (!io.Fonts->IsBuilt()) {
    io.Fonts->Build();
}
```

**Comment by ocornut** [comment]
See https://github.com/ocornut/imgui/blob/master/docs/FONTS.md#troubleshooting
"Font atlas texture fails to upload to GPU."

TL;DR: with current system its not really wise to ever use e.g. `GetGlyphRangesChineseFull()`.
If your application has localized text you may use a GlyphRangerBuilder to build only character necessary by localization.

We should rework the system and address this this year.

**Comment by ocornut** [comment]
FYI, work on `dynamic_fonts` branch aimed to be in merged in 1.92 (#8465) solves this.

**Comment by DubbleClick** [comment]
That's really cool, we had already solved the issue by replacing the font atlas, but this will simplify and enable custom sizes for users more easily. We will give it a shot over the next weeks and report back in case we find any issues.
